Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift

```diff
@@ -0,0 +1,287 @@
+    case let .tail(closure, unexpected):
+      closure.trimmed.write(to: &text)
+      unexpected.write(to: &text)
+    }
+    guard
+      let object = try? JSONSerialization.jsonObject(with: text.data(using: .utf8)!),
+      let serial = object as? Dictionary<String, Any>
+    else {
+      return nil
+    }
+
+    return self.build(from: serial)
+  }
+}
+
+extension ConvertJSONToCodableStructRefactor {
+  public enum Preflight {
+    case closure(ClosureExprSyntax)
+    case tail(ClosureExprSyntax, UnexpectedNodesSyntax)
+  }
+  public static func preflightRefactoring(_ closure: ClosureExprSyntax) -> Preflight? {
+    if let file = closure.parent?.parent?.parent?.as(SourceFileSyntax.self),
+      let unexpected = file.unexpectedBetweenStatementsAndEndOfFileToken
+    {
+      return .tail(closure, unexpected)
+    }
+
+    if closure.hasError,
+      closure.unexpectedBetweenStatementsAndRightBrace != nil
+    {
+      return .closure(closure)
+    }
+    return nil
+  }
+}
+
+extension ConvertJSONToCodableStructRefactor {
+  private static func build(from jsonDictionary: Dictionary<String, Any>) -> DeclSyntax {
+    return
+      """
+      \(raw: self.buildStruct(from: jsonDictionary))
+      """
+  }
+
+  private static func buildStruct(
+    named name: String = "JSONValue",
+    at depth: Int = 0,
+    from jsonDictionary: Dictionary<String, Any>
+  ) -> String {
+    let members = self.buildStructMembers(at: depth + 1, from: jsonDictionary)
+    return
+      """
+      \(String(repeating: " ", count: depth * 2))struct \(name): Codable {
+      \(members.joined(separator: "\n"))
+      \(String(repeating: " ", count: depth * 2))}
+      """
+  }
+
+  private static func buildStructMembers(
+    at depth: Int,
+    from jsonDictionary: Dictionary<String, Any>
+  ) -> [String] {
+    var members = [String]()
+
+    var uniquer = Set<String>()
+    var nestedTypes = [(String, Dictionary<String, Any>)]()
+    func addNestedType(_ name: String, object: Dictionary<String, Any>) {
+      guard uniquer.insert(name).inserted else {
+        return
+      }
+      nestedTypes.append((name, object))
+    }
+
+    for key in jsonDictionary.keys.sorted() {
+      guard let value = jsonDictionary[key] else {
+        continue
+      }
+
+      let type = self.jsonType(of: value, suggestion: key)
+      if case let .object(name) = type, let subObject = value as? Dictionary<String, Any> {
+        addNestedType(name, object: subObject)
+      } else if case let .array(innerType) = type,
+        let array = value as? [Any],
+        let (nesting, elementType) = innerType.outermostObject
+      {
+        if let subObject = array.unwrap(nesting) {
+          addNestedType(elementType, object: subObject)
+        }
+      }
+
+      let member = "\(String(repeating: " ", count: depth * 2))var \(key): \(type.stringValue)"
+      members.append(member)
+    }
+
+    if !nestedTypes.isEmpty {
+      members.append("")
```

Comment:
Having to append an empty string to members to (presumably) bet an empty line between the members seems wrong. I think we should add an empty line between all the members instead.